executor/metric_reader.go

```diff
@@ -193,3 +194,53 @@ func (c *queryClient) URL(ep string, args map[string]string) *url.URL {
 	ep = strings.Replace(ep, "api/v1", "pd/api/v1/metric", 1)
 	return c.Client.URL(ep, args)
 }
+
+// MetricSummaryRetriever uses to read metric data.
+type MetricSummaryRetriever struct {
+	dummyCloser
+	table     *model.TableInfo
+	extractor *plannercore.MetricTableExtractor
+	quantiles []float64
+	retrieved bool
+}
+
+func (e *MetricSummaryRetriever) retrieve(ctx context.Context, sctx sessionctx.Context) ([][]types.Datum, error) {
+	if e.retrieved || e.extractor.SkipRequest {
+		return nil, nil
+	}
+	e.retrieved = true
+	totalRows := make([][]types.Datum, 0, len(infoschema.MetricTableMap))
+	e.quantiles = []float64{1, 0.999, 0.99, 0.90, 0.80}
+	tps := make([]*types.FieldType, 0, len(e.table.Columns))
+	for _, col := range e.table.Columns {
+		tps = append(tps, &col.FieldType)
+	}
+	startTime := e.extractor.StartTime.Format(plannercore.MetricTableTimeFormat)
+	endTime := e.extractor.EndTime.Format(plannercore.MetricTableTimeFormat)
+	for name, def := range infoschema.MetricTableMap {
+		sqls := e.genMetricQuerySQLS(name, startTime, endTime, def.Quantile)
+		for _, sql := range sqls {
+			rows, _, err := sctx.(sqlexec.RestrictedSQLExecutor).ExecRestrictedSQL(sql)
```

Comment:
According to your manual test, does it need to retry after network failure?

Reply:
When querying the single metric tables, there is already has the retry internal.